rfcs/0071-target-json-parser.md

```diff
@@ -0,0 +1,172 @@
+    "libs": ["cudnn"],
+    "target_host": {
+        "id": "llvm",
+        "system_lib": True,
+        "mtriple": "aarch64-linux-gnu",
+        "mattr": "+neon"
+    }
+}
+```
+
+We can see that there are additional fields which are of interest to TVM, note-ably `keys` and `libs` which we currently do not apply parsing to on `Target` instantiation. Extending the `TargetKind` `preprocessor` beyond `attrs` enables to customise parsing of the entire `Target`, enabling the values passed by the user to be used to infer other properties used during compilation.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Alongside the existing `set_attrs_preprocessor` method on `TargetKind`, there will be an alternative `set_target_parser` method, this new parser will take precedence over the `attrs` preprocessor if present:
+
+```c++
+TVM_REGISTER_TARGET_KIND("target", kDLCPU)
+    .set_target_parser(TargetParser);
+```
+
+The canonical JSON form of `Target` will be passed to the new `Target` parser and the parser will return the transformed variant in JSON form for further steps:
+
+```c++
+using TargetJSON = Map<String, ObjectRef>;
+TargetJSON TargetParser(TargetJSON target) {
+    // ... transforms ...
+    return target;
+}
+```
+
+The parser will have to be capable of handling the diversity of types of `Target` in TVM, therefore the underlying mechanism of the parser is left as an implementation detail. Using the example of pre-processing the `keys` attribute (used for detecting appropriate schedules), it can be seen how this can apply to various `Target`s. 
+
+## TVM Target's Directly Mapping to a Backend's Target
+Take the example of pre-processing `keys` (in this case using the `cuda` `Target`):
+```c++
+using TargetJSON = Map<String, ObjectRef>;
+
+TargetJSON CUDAParser(TargetJSON target) {
+    if (IsSuper(target)) {
+        target["keys"].push_back("super_cuda");
+    }
+}
+
+TVM_REGISTER_TARGET_KIND("cuda", kDLGPU)
+    .set_target_parser(CUDAParser);
+```
+
+This takes the `attrs` from `Target` and maps them to relevant `keys` for use when selecting schedules:
+
+```c++
+Target my_target("cuda -msuper");
+my_target->keys; // ["cuda", "gpu", "super_cuda"] <-- "cpu" and "cuda" are taken from default keys - "super_cuda" is added
+```
+
+## TVM Target's Mapping to a Backend with Multiple Target's
+The previous example would work for `Target`s which map to a specific architecture, such as `cuda`. To parse a `Target` which has a number of its own targets, such as `llvm`, the parser can be broken down within the parent parser:
+
+```c++
+using TargetJSON = Map<String, ObjectRef>;
+
+TargetJSON AArch64TargetParser(TargetJSON target) {
+    target["keys"].push_back("arm_cpu");
+    return target;
+}
+
+TargetJSON x86TargetParser(TargetJSON target) {
+    target["keys"].push_back("x86_64");
+    return target;
+}
+
+TargetJSON CPUTargetParser(TargetJSON target) {
+    if (IsAArch64Target(target)) {
```

Comment:
Unless the details of the "kind" are listed in the JSON, there is no way of knowing what architecture the target refers to.